subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/install/internal/AdoptOpenJdkRemoteBinary.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.jvm.toolchain.install.internal;
+
+import com.google.common.io.Files;
+import net.rubygrapefruit.platform.SystemInfo;
+import org.apache.commons.io.FileUtils;
+import org.gradle.api.GradleException;
+import org.gradle.api.JavaVersion;
+import org.gradle.internal.os.OperatingSystem;
+import org.gradle.jvm.toolchain.JavaToolchainSpec;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+public class AdoptOpenJdkRemoteBinary {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AdoptOpenJdkRemoteBinary.class);
+
+    private final SystemInfo systemInfo;
+    private final OperatingSystem operatingSystem;
+
+    @Inject
+    public AdoptOpenJdkRemoteBinary(SystemInfo systemInfo, OperatingSystem operatingSystem) {
+        this.systemInfo = systemInfo;
+        this.operatingSystem = operatingSystem;
+    }
+
+    // TODO: [bm] this might be downloading the same jdk multiple times in parallel builds
+    // TODO: [bm] Alternative is using RepositoryTransportFactory+ExternalResourceAccessor+ChecksumService
```

Comment:
How hard is it to reuse the existing infrastructure? We might get things like progress logging (see comment below) out of the box and maybe even a build operation that shows up in build scans.

If not addressed here, can you create a follow up for this. I think this is important.

Reply:
I can push a follow PR with those changes (I have them locally and that was my first approach). It has some additional complexities that might require a second brain to resolve. If it works for you, I'd try and separate that in an additional PR. My goal is that all `TODO: [bm]` are resolve before the issue closes

Reply:
👍 Fine with me. Let's address this topic in a follow up.